src/twisted/web/static.py

```diff
@@ -271,9 +271,31 @@ def ignoreExt(self, ext):
     childNotFound = resource.NoResource("File not found.")
     forbidden = resource.ForbiddenResource()
 
+
     def directoryListing(self):
-        return DirectoryLister(self.path,
-                               self.listNames(),
+        """
+        Return a resource that generates an HTML listing of the
+        directory this path represents.
+
+        @return: A L{resource.Resource} that renders the directory to
+            HTML.
+        """
+        if _PY3:
+            path = self.path
+            names = self.listNames()
+        else:
+            # DirectoryLister works in terms of native strings, so on
```

Comment:
Do we not need to do the inverse (`.asTextMode()`) on PY3 then?

Reply:
No, but for a bad reason: `File` and `DirectoryLister` are tricky.

[`getChild` now decodes all descendant paths to text](https://github.com/twisted/twisted/blob/twisted-18.4.0/src/twisted/web/static.py#L301-L309), so that by the time the empty path is reached and `directoryListing` called, the path given to `DirectoryLister` is guaranteed to be text, _unless_ of course the path being listed is the `File`'s _own_ path.

Given:

```python
from twisted.web import server, static
from twisted.internet import defer, endpoints, task


@task.react
def main(reactor):
    site = server.Site(static.File(b"/path/to/some/directory/"))
    listener = endpoints.serverFromString(reactor, "tcp:8080")
    d = listener.listen(site)

    @d.addCallback
    def _(_):
        return defer.Deferred()

    return d
```

Assuming that `/path/to/some/directory/child` exists, accessing `http://localhost:8080/child/` or any other descendant creates a `DirectoryLister` with a text path.  Accessing `http://localhost:8080/`, however, creates one with a _bytes_ path.  

Surprise! On Python 3, this is fine.  That's `DirectoryLister` _itself_  [decodes bytes descendants to text](https://github.com/twisted/twisted/blob/twisted-18.4.0/src/twisted/web/static.py#L1004-L1006) on Python 3.

So the _actual_ admitted by `DirectoryLister` is:

| `DirectoryLister` parameter  | Python 3         | Python 2        |
|------------------------------|------------------|-----------------|
| pathname                     | `str` or `bytes` | `str`           |
| dirs                         | `str` or `bytes` | `str`           |
| contentTypes                 | `dict` of `str`  | `dict` of `str` |
| contentEncodings             | `dict` of `str`  | `dict` of `str` |
| defaultType                  | `str`            | str             |

Which is not exactly what I said in the commit message, but does mean this part of the PR is correct (I think.)

Reply:
Every day I'm a little more sympathetic towards web2's abandonment of this API :).